New set of classes and sub-classes created for Fluid and Particle properties.

npannozzo opened the pull request:
Hi @manuGil,

Here I pushed the set of classes created for the 'medium' (which I have eventually called 'fluid') and the 'particle' properties. Could you please revise the style and structure of the code and let me know if there is anything missing or incorrect? If the code looks good, please don't merge it yet, because I would still like the rest of the team to have a look and check that there is no content missing before merging. In the end, I realized that we didn't need subclasses for water and wind, but just one general fluid class was enough, as the only difference between the calculation of the wind and water properties is the different fluid density, which can defined manually by the user in the configuration file anyway so, if we give the fluid density parameter a general name in the configuration file (e.g., rhoFluid) and we manually let it set by the user (based on the type of fluid used), then there is no need to separate the different types of fluids in different sub-classes after all, for the fluid properties calculations. If we make rhoFluid manually settable, we also make our life easier with the definition of the different sediment (particle) classes, because most of the sediment properties calculations also depend on that parameter and, if we defined default different rhos for different types of fluids, we would then end up needing an endless number of subclasses for the calculations of the sediment properties.

Many thanks for your help with this! I learned so much from your lecture!

Natascia

npannozzo committed:
New set of classed created for medium properties.

npannozzo committed:
Classes and sub-classes for Fluid and Particle.

manuGil commented:
@npannozzo I am glad you realize this yourself: " if we defined default different rhos for different types of fluids, we would then end up needing an endless number of subclasses for the calculations of the sediment properties." That is a well-know disadvantage of over using inheritance.

manuGil commented on src/sedtrails/transport_converter/plugins/physics/library/Fluid.py:
```diff
@@ -0,0 +1,40 @@
+@dataclass
+class Fluid:
+    """
+    For example, water and wind.
```
I suggest you use a description such as: "Represents a fluid in a simulation. For example, water or wind"

manuGil commented on src/sedtrails/transport_converter/plugins/physics/library/Particle.py:
```diff
@@ -0,0 +1,249 @@
+        """
+        pass
+
+    def Dstar_Background(self) -> float:
```
if all you intend every method to be abstract (given you did no define what they do here). you have to use `@abstractmethod` on top of each of them

manuGil commented on src/sedtrails/transport_converter/plugins/physics/library/Particle.py:
```diff
@@ -0,0 +1,249 @@
+        """
+        return S.dTracer / S.dBackground
+
+class Sand(Sediment(Particle)):
```
Use `class Sand(Sediment)` instead, `Sediment` already inherits from Particle, and therefore, you can access its methods by inheriting from Sediment alone.

npannozzo committed:
Modified Particle classes based on discussion
regarding internal data structure and added
more extensive comments in both files.

npannozzo commented:
Hello, I have modified the Particle classes file following the discussion regarding the internal data structure that we had last week. I have also added more extensive descriptions to both files. I still have a doubt about whether we should have a class for "Passive(Particle)" in this file because it has no physical properties. Please, let me know if there any further improvements to make. Thanks for your help with this!

monica-aguil committed:
Update Particle.py

manuGil commented on src/sedtrails/transport_converter/plugins/physics/library/Fluid.py:
```diff
@@ -0,0 +1,57 @@
+    """
+    
+    # concrete method
+    def abstau_mean(self, D) -> float:
```
Please, give a more descriptive name to D... and also add a docstring and type hint (if we know that already). Also do the same in the cases below

manuGil commented on src/sedtrails/transport_converter/plugins/physics/library/Fluid.py:
```diff
@@ -0,0 +1,57 @@
+        """
+        return np.abs(D.tau_mean)
+    
+    def abstau_max(self, D) -> float:
+        """
+        Calculates the absolute maximum shear stress of the fluid [N/m^2].
+        """
+        return np.abs(D.tau_max)
+    
+    def U_mag(self, Ux, Uy) -> float:
+        """
+        Calculates the total velocity magnitude of the fluid [m/s].
+        """
+        return np.sqrt(Ux ** 2 + Uy ** 2)
+    
+    def ustar_mean(self, abstau_mean, S) -> float:
+        """
+        Calculates the mean friction velocity of the fluid [m/s].
+        """
+        return np.sqrt(abstau_mean / S.rhoFluid)
+
+    def ustar_max(self, abstau_max, S) -> float:
+        """
+        Calculates the maximum friction velocity of the fluid [m/s].
+        """
+        return np.sqrt(abstau_max / S.rhoFluid)
```
I see that in your docstrings you defined class attributes, but aren't present in the dataclass itself. The purpose of a `dataclass` is to hold data during computation. From what I can gather from your methods, some of those are computing the values for the attribute classes. In which cases, what they do is to update the attribute values. 

For example:
```python
@dataclass
class Fluid:
""" Your dockstirngs"""
    mean_tau_max: float = None # assuming the class can be initialized with a None value for this
    # other attributes
    
    # then the methods update those values
    def abstau_max (self, D) -> float:
        self.mean_tau_max = np.abs(D.tau_max)
        return self.mean_tau_max # you can return the value if that's the goal
    ....

 ```
 But, that will be the way to go if I have understood your intentions correctly. Let's discuss this before you make changes.

npannozzo replied:
Thanks for the feedback, Manuel! Let's discuss this over videocall on Wednesday 23rd.

manuGil commented on src/sedtrails/transport_converter/plugins/physics/library/Particle.py:
This is similar to #110 We should consolidate this.

manuGil replied:
This is has been address in #171

manuGil reviewed (changes requested):
Thanks for this Natscia, 
There are a few things I think we should solve together. Let me know when we can look at my comments together.

manuGil commented:
This contribution was moved to #171. Closing this without merge
